File: solutions/example_solution.clj
(ns solution
  (:require [ring.middleware.reload :refer [wrap-reload]]
            [ring.util.response :refer [response status]]
            [compojure.core :refer [GET POST defroutes]]
            [clojure.java.jdbc :as jdbc]
            [clojure.string :as str]
            [cheshire.core :as json]
            [clojure.pprint :refer [pprint]]
            [ring.adapter.jetty :refer [run-jetty]]
            [ring.middleware.json :refer [wrap-json-response wrap-json-params]]))


(def db-spec  {:classname   "org.sqlite.JDBC"
               :subprotocol "sqlite"
               :subname     "../database.db"})

(defn messages []
  (let [messages  (jdbc/query db-spec ["SELECT * FROM messages"])
        state-raw (jdbc/query db-spec ["SELECT id, value FROM state"])
        state     (reduce (fn [acc val] (assoc acc (:id val) (:value val))) {} state-raw)]
    (for [msg messages
          :let [m (:body msg)]]
      (str/replace m #"\{(.+?)\|(.*?)\}"
                   (fn [[_ id fallback]]
                     (get state id fallback))))))

(defn search [query]
  (letfn   [(extract-text [block]
              (if (sequential? block)
                (mapcat extract-text block)
                (map (fn [[key val]]
                       (if (sequential? val)
                         (str/join (mapcat extract-text val))
                         (when (not= :type key)
                           val)))
                     (seq block))))]

    (let [query-terms (str/split (str/lower-case query) #"\s+")
          answers-raw (jdbc/query db-spec ["select
                                              a.id, a.title, b.content
                                            from
                                              answers a
                                            join blocks b
                                              on a.id=b.answer_id"])
          answers     (map (fn [{:keys [id title content]}]
                             {:id      id
                              :title   title
                              :content (json/parse-string content keyword)})
                           answers-raw)]

      (filter (fn [answer]
                (let [content-text (str/join (extract-text (:content answer)))
                      full-text (str/lower-case (str (:title answer) " " content-text))]
                  (every? true?
                          (map (fn [term] (str/includes? full-text term)) query-terms))))
              answers))))

(defroutes app-routes
  (GET "/messages" [] (response (messages)))
  (POST "/search" [query]
        (if (and (string? query) (not= "" query))
          (response (search query))
          (-> (response "Query missing or empty") (status 400)))))

(defn app []
  (-> #'app-routes
      (wrap-json-response)
      (wrap-json-params)
      (wrap-reload {:dirs "."})))

(defn -main [& args]
  (run-jetty (app) {:port 5000 :join? false}))

File: solutions/example_solution.js
const sqlite3 = require('sqlite3').verbose();
var express = require('express')

var app = express()
app.use(express.json());

let db = new sqlite3.Database('../database.db');

// Return all messages
app.get('/messages', function (req, res) {
    db.all("select body from messages",{},(err, messages) => {
        db.all("select id, value from state",{},(err, state_raw) => {
            // put it into a dictionary for easier handling
            let state = state_raw.reduce((acc, val) => {
                acc[val.id] = val.value
                return acc
            }, {})

            // Replace all matches
            let fullMessages = messages.map(message => {
                return message.body.replace(/{(.+?)\|(.*?)}/g, (full, id, fallback) => {
                    return state[id] || fallback
                });
            });
            res.send(fullMessages)
        });
    });
})

// Search for answers
app.post('/search', function (req, res) {
    let query = req.body.query
    if(! query)
        return res.status(400).send("Bad query")

    let queryTerms = query.toLowerCase().split(" ")

    db.all(`select
                a.id, a.title, b.content
            from
                answers a
            join blocks b
                on a.id=b.answer_id`, {}, (err, rows_raw) => {

               let rows = rows_raw.map((r) => {
                   return {"id": r.id, "title": r.title, content: JSON.parse(r.content)}
               })

               let matched_rows = rows.filter(answer => {

                   // create a big string that includes all content text
                   let extract_text = (block) => {
                       if (Array.isArray(block)) {
                           return block.map(extract_text)
                       }
                       var text = ""
                       for (let [key, value] of Object.entries(block)) {
                           if(Array.isArray(value)){
                               text += value.map(extract_text)
                           }
                           else if(key != "type") {
                               text += " " +value
                           }
                       }
                       return text
                   }

                   // mash it all together and normalize it
                   let fulltext = (answer.title + " " + extract_text(answer.content).join(" ")).toLowerCase()

                   // see if all the terms show up
                   for(term of queryTerms) {
                       if(! fulltext.includes(term))
                           return false
                   }
                   return true
               })

               res.status(200).send(matched_rows)
           });

})

var server = app.listen(5000, function() {
    console.log('Express server listening on port ' + server.address().port);
});
